Pass ConfigSpace definition file directly to BOHB

khituras opened the pull request:
This PR show a possible way to pass a path to a ConfigSpace file to BOHB so that it can read the file directly.
Thus, the translation from the NNI search space to the ConfigSpace format is circumvented.
Here, I introduced the special search space parameter named `ConfigSpace` that whose value is the path to the ConfigSpace file. `bohb_advisor.py` checks for this parameter and loads the file from the given path, if present.
The modification in `launcher.py` extends a relative ConfigSpace definition file path to an absolute path, if necessary.

I do not expect this PR to be accepted as-is. It is rather here for reference for the issue #4140 where I requested a similar functionality as a new feature in NNI.

ultmaster commented:
As I mentioned in the comments, my suggestion is that the pcs file should be passed as an init argument.

For how to support an optional argument, please see: https://nni.readthedocs.io/en/stable/Tutorial/InstallCustomizedAlgos.html#optional-create-a-validator-to-validate-classargs

khituras commented:
I did add an optional classArgs parameter named `config_space`. To make it optional, I had to add a method to `ClassArgsValidator` because there previously was no method to check for arbitrary strings. But we need this since we want to be able to pass a file name or a path. Thus, I added the `keyType()` method that just checks if the parameter has the right type, and added a check to the `bohb_advisor` for `config_space` being a string.

Now you can specify the ConfigSpace file like so (note the new `config_space` property at the end):

```yaml
advisor:
  builtinAdvisorName: BOHB
  classArgs:
    optimize_mode: maximize
    min_budget: 1
    max_budget: 9
    eta: 3
    min_points_in_model: 7
    top_n_percent: 15
    num_samples: 64
    random_fraction: 0.33
    bandwidth_factor: 3.0
    min_bandwidth: 0.001
    config_space: bohb-parameters.pcs
```

When doing this, the contents of the actual search space file are unused. The file can thus just contain `{}` to be valid JSON.

If this is acceptable to you, I could add the respective documentation to the BOHB docs.

ultmaster commented on nni/utils.py:
```diff
@@ -305,3 +305,6 @@ def range(self, key, keyType, start, end):
             And(lambda n: start <= n <= end, error='%s should be in range of (%s, %s)!' % (key, start, end))
         )
+
+    def keyType(self, key, keyType):
```
```python
def path(self, key, keyType):
   # check whether this path is valid and exists.
```

khituras replied:
Will make this change.

ultmaster commented on nni/tools/nnictl/launcher.py:
```diff
@@ -421,6 +421,18 @@ def launch_experiment(args, experiment_config, mode, experiment_id, config_versi
+    # ConfigSpace format by defining the parameter 'ConfigSpace' as a 'choice' parameter
+    # with '_value' being the path to the configuration file. If the path is relative, we
+    # need to expand it
+    if experiment_config.get('advisor') and experiment_config['advisor'].get('name') == 'BOHB'\
```
Why do we still need this?

khituras replied:
To make the relative path to the ConfigSpace file an absolute one. With this code, the configSpace file can just reside in the `trialCodeDirectory` and be referenced by name. Without it, one would have to define the complete path.
The comment is deprecated, though, will change it.

khituras replied:
I have now moved that part to `launcher_utils.py` because this is where other paths are handled as well.

ultmaster reviewed (commented):
Design looks good. Might need to resolve another few detailed comments.

ultmaster reviewed (approved):
LGTM. Zhe please take a look. :)

liuzhe-lz commented on nni/algorithms/hpo/bohb_advisor/bohb_advisor.py:
```diff
@@ -468,48 +471,56 @@ def handle_update_search_space(self, data):
+        cs = None
+        logger.debug(f'Received data: {data}')
+        if self.config_space:
+            logger.debug(f'Got a ConfigSpace file path, parsing the search space directly from {self.config_space}.\
```
Suggest raise log level to info.

ultmaster commented:
Reopen to re-trigger pipeline.
